pkg/collector/stats/types/types.go

```diff
@@ -38,90 +46,92 @@ func (s *UInt64Stat) SetNewDeltaValue(newDelta uint64, sum bool) error {
 		return fmt.Errorf("the aggregated value has overflowed")
 	}
 	if sum {
 		// sum is used to accumulate metrics from different processes
-		s.Delta += newDelta
+		s.delta.Add(newDelta)
 	} else {
-		s.Delta = newDelta
+		if swapped := s.delta.CompareAndSwap(s.delta.Load(), newDelta); !swapped {
+			return fmt.Errorf("the delta value has been changed by another process")
+		}
 	}
-	s.Aggr += newDelta
+	// update aggregated value
+	s.aggr.Add(newDelta)
 	return nil
 }
 
 // SetNewAggr set new read aggregated value (e.g., from cgroup, energy files)
 func (s *UInt64Stat) SetNewAggr(newAggr uint64) error {
-	if newAggr == 0 || newAggr == s.Aggr {
+	currAggr := s.aggr.Load()
+	if newAggr == 0 || newAggr == currAggr {
 		// if a counter has not changed, we skip it
 		return nil
 	}
 	// verify aggregated value overflow
 	if newAggr == math.MaxUint64 {
 		// we must set the the value to 0 when overflow, so that prometheus will handle it
-		s.Aggr = 0
+		s.aggr.Store(0)
 		return fmt.Errorf("the aggregated value has overflowed")
 	}
-	if (s.Aggr > 0) && (newAggr > s.Aggr) {
-		s.Delta = newAggr - s.Aggr
+	if (currAggr > 0) && (newAggr > currAggr) {
+		if swapped := s.delta.CompareAndSwap(s.delta.Load(), newAggr-currAggr); !swapped {
+			return fmt.Errorf("the delta value has been changed by another process")
+		}
+	}
+	if swapped := s.aggr.CompareAndSwap(currAggr, newAggr); !swapped {
+		return fmt.Errorf("the aggregated value has been changed by another process")
 	}
-	s.Aggr = newAggr
 	return nil
 }
 
 func (s *UInt64Stat) GetDelta() uint64 {
-	return s.Delta
+	return s.delta.Load()
 }
 
 func (s *UInt64Stat) GetAggr() uint64 {
-	return s.Aggr
+	return s.aggr.Load()
 }
 
-func NewUInt64StatCollection() *UInt64StatCollection {
-	return &UInt64StatCollection{
-		Stat: make(map[string]*UInt64Stat),
-	}
+func NewUInt64StatCollection() UInt64StatCollection {
+	return make(map[string]*UInt64Stat)
 }
 
-// UInt64StatCollection keeps a collection of UInt64Stat
-type UInt64StatCollection struct {
-	Stat map[string]*UInt64Stat
-}
+type UInt64StatCollection map[string]*UInt64Stat
 
-func (s *UInt64StatCollection) SetAggrStat(key string, newAggr uint64) {
-	if instance, found := s.Stat[key]; !found {
-		s.Stat[key] = &UInt64Stat{
-			Aggr:  newAggr,
-			Delta: 0,
-		}
+func (s UInt64StatCollection) SetAggrStat(key string, newAggr uint64) {
+	if instance, found := s[key]; !found {
+		stat := &UInt64Stat{}
+		stat.aggr.Store(newAggr)
+		s[key] = stat
 	} else {
 		if err := instance.SetNewAggr(newAggr); err != nil {
 			klog.V(3).Infoln(err)
 		}
 	}
 }
 
-func (s *UInt64StatCollection) AddDeltaStat(key string, newDelta uint64) {
-	if instance, found := s.Stat[key]; !found {
-		s.Stat[key] = &UInt64Stat{
-			Aggr:  newDelta,
-			Delta: newDelta,
-		}
+func (s UInt64StatCollection) AddDeltaStat(key string, newDelta uint64) {
+	if instance, found := s[key]; !found {
+		stat := &UInt64Stat{}
+		stat.aggr.Store(newDelta)
+		stat.delta.Store(newDelta)
+		s[key] = stat
```

Comment:
How about we use the constructors / NewX instead of handling the internal datatypes (now not exported) ? 

```suggestion
		s[key] = NewUnit64Stat(newDelta, newDelta)
```